src/cio_websocket.c

```diff
@@ -0,0 +1,466 @@
+	}
+
+	if (unlikely((ws->ws_flags.fin == 0) && (ws->ws_flags.opcode >= CIO_WEBSOCKET_CLOSE_FRAME))) {
+		// TODO: handle_error(s, WS_CLOSE_PROTOCOL_ERROR);
+		return;
+	}
+
+	if (ws->ws_flags.fin == 0) {
+		if (ws->ws_flags.opcode != CIO_WEBSOCKET_CONTINUATION_FRAME) {
+			if (unlikely(ws->ws_flags.is_fragmented)) {
+				// TODO: log_err("Overwriting Opcode of unfinished fragmentation!");
+				// TODO: handle_error(s, WS_CLOSE_PROTOCOL_ERROR);
+				return;
+			}
+
+			ws->ws_flags.is_fragmented = 1;
+			ws->ws_flags.frag_opcode = ws->ws_flags.opcode;
+			ws->ws_flags.opcode = CIO_WEBSOCKET_CONTINUATION_FRAME;
+		} else {
+			if (unlikely(ws->ws_flags.rsv != 0)) {
+				// TODO:log_err("RSV-Bits must be 0 during continuation!");
+				// TODO:handle_error(s, WS_CLOSE_PROTOCOL_ERROR);
+				return;
+			}
+
+			if (unlikely(!(ws->ws_flags.is_fragmented))) {
+				//TODO: log_err("No start frame was send!");
+				//TODO: handle_error(s, WS_CLOSE_PROTOCOL_ERROR);
+				return;
+			}
+		}
+	}
+
+	if (unlikely(ws->ws_flags.is_fragmented && ((ws->ws_flags.opcode < CIO_WEBSOCKET_CLOSE_FRAME) && (ws->ws_flags.opcode > CIO_WEBSOCKET_CONTINUATION_FRAME)))) {
+		// log_err("Opcode during fragmentation must be 0x0!");
+		// handle_error(s, WS_CLOSE_PROTOCOL_ERROR);
+		return;
+	}
+
+	bool last_frame = false;
+	unsigned int opcode;
+	if (ws->ws_flags.opcode == CIO_WEBSOCKET_CONTINUATION_FRAME) {
+		opcode = ws->ws_flags.frag_opcode;
+		if (unlikely(ws->ws_flags.fin ==1)) {
+			last_frame = true;
+		}
+	} else {
+		opcode = ws->ws_flags.opcode;
+	}
+
+	switch (opcode) {
+	case CIO_WEBSOCKET_BINARY_FRAME:
+		if (likely(ws->onbinaryframe_received != NULL)) {
+			ws->onbinaryframe_received(ws, data, length, last_frame);
+		} else {
+			// TODO: handle_error(s, WS_CLOSE_UNSUPPORTED);
+		}
+
+		break;
+
+	case CIO_WEBSOCKET_TEXT_FRAME:
+		if (likely(ws->on_textframe_received != NULL)) {
+			ws->on_textframe_received(ws, (char *)data, length, last_frame);
+		} else {
+			// TODO: handle_error(s, WS_CLOSE_UNSUPPORTED);
+		}
+
+		break;
+
+	case CIO_WEBSOCKET_PING_FRAME:
+		if (unlikely(length > WS_SMALL_FRAME_SIZE)) {
+			//TODO: handle_error(s, WS_CLOSE_PROTOCOL_ERROR);
+		} else {
+			// TODO: send_pong_frame(ws, data, length);
+			if (ws->on_ping != NULL) {
+				ws->on_ping(ws, data, length);
+			}
+		}
+
+		break;
+
+	case CIO_WEBSOCKET_PONG_FRAME:
+		if (unlikely(length > WS_SMALL_FRAME_SIZE)) {
+			//TODO: handle_error(s, WS_CLOSE_PROTOCOL_ERROR);
+		} else {
+			if (ws->on_pong != NULL) {
+				ws->on_pong(ws, data, length);
+			}
+		}
+
+		break;
+
+	case CIO_WEBSOCKET_CLOSE_FRAME: {
+		uint16_t status_code = CIO_WEBSOCKET_CLOSE_NORMAL;
+		if (length >= 2) {
+			memcpy(&status_code, data, sizeof(status_code));
+			status_code = cio_be16toh(status_code);
+		}
+
+		if (length > 2) {
+			// TODO: struct cjet_utf8_checker c;
+			// TODO: cjet_init_checker(&c);
+			// TODO: if (!cjet_is_byte_sequence_valid(&c, frame + 2, length - 2, true)) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Remove this block of commented out code. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=c%3ACommentedCode)
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this "TODO" comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=c%3AS1135)